File: src/Core/Type.cpp
//===-- Core/Type.cpp ---------------------------------------------------------------*- C++ -*-===//
//
//                                    S E R I A L B O X
//
// This file is distributed under terms of BSD license.
// See LICENSE.txt for more information
//
//===------------------------------------------------------------------------------------------===//
//
/// \file
/// This file contains definitions for types recognized by Serialbox.
///
//===------------------------------------------------------------------------------------------===//

#include "serialbox/Core/Type.h"
#include "serialbox/Core/Unreachable.h"

namespace serialbox {

std::string TypeUtil::toString(TypeID id) {
  switch(id) {
  case TypeID::Boolean:
    return std::string("bool");
  case TypeID::Int32:
    return std::string("int");
  case TypeID::Int64:
    return std::string("int64");
  case TypeID::Float32:
    return std::string("float");
  case TypeID::Float64:
    return std::string("double");
  case TypeID::String:
    return std::string("std::string");
  case TypeID::Invalid:
    return std::string("<invalid>");
  }
}

std::ostream& operator<<(std::ostream& stream, const OpenModeKind& mode) {
  switch(mode) {
  case OpenModeKind::Write:
    return (stream << std::string("Write"));
  case OpenModeKind::Read:
    return (stream << std::string("Read"));
  case OpenModeKind::Append:
    return (stream << std::string("Append"));
  }
}

std::ostream& operator<<(std::ostream& stream, const TypeID& t) {
  return (stream << TypeUtil::toString(t));
}

int TypeUtil::sizeOf(TypeID id) noexcept {
  switch(id) {
  case TypeID::Boolean:
    return 1;
  case TypeID::Float32:
  case TypeID::Int32:
    return 4;
  case TypeID::Int64:
  case TypeID::Float64:
    return 8;
  default:
    serialbox_unreachable("invalid TypeID for TypeUtil::sizeOf");
  }
}

} // namespace serialbox

File: src/serialbox-python/serialbox/type.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
##===-----------------------------------------------------------------------------*- Python -*-===##
##
##                                   S E R I A L B O X
##
## This file is distributed under terms of BSD license.
## See LICENSE.txt for more information.
##
##===------------------------------------------------------------------------------------------===##
##
## This file contains the type and enum definitions of Serialbox
##
##===------------------------------------------------------------------------------------------===##

from enum import Enum

import numpy as np


class TypeID(Enum):
    """Type-id of types recognized by serialbox.

    This enum corresponds to the enum definitions of the C/C++ library.
    """
    Invalid = 0

    Boolean = 1
    Int32 = 2
    Int64 = 3
    Float32 = 4
    Float64 = 5
    String = 6

    Array = 0x10
    ArrayOfBoolean = Array | Boolean
    ArrayOfInt32 = Array | Int32
    ArrayOfInt64 = Array | Int64
    ArrayOfFloat32 = Array | Float32
    ArrayOfFloat64 = Array | Float64
    ArrayOfString = Array | String


BooleanTypes = (bool, np.bool,)
Int32Types = (int, np.int8, np.int16, np.int32, np.uint8, np.uint16, np.uint32,)
Int64Types = (np.int64, np.uint64,)
Float32Types = (np.float32,)
Float64Types = (float, np.float64,)
StringTypes = (bytes, str,)

__all__ = ['TypeID', 'BooleanTypes', 'Int32Types', 'Int64Types', 'Float32Types', 'Float64Types',
           'StringTypes']
